### PR TITLE
Econ

### PR DESCRIPTION
**Pull request by gawquon**
Rough draft of the EconomyManager. Could use some eyes to reign in my tendency for over-convolutedness.
Current plan goes something like this:

1. For each centralized country get a CP budget based on population and culture TechGroup
2. Read in terrain based effect Matrices
3. For each substate in a centralized country get a CP budget based on population and terrain/state modifiers
4. Figure out the "national budget" of each country based on eurocentrism, colonialism, GP status, EU4 gov reform, manufactories maybe?
4b. National Budget is like 30% Agriculture, 25% Industry, etc.
4c. Excludes Military, Navy, government administration, ports and rails
5. Each Substate spends it's CP budget according to it's owners national budget and local terrain/state modifiers
5b. When a substate goes over its infrastructure, try to build ports/rails. If you can't, go over the limit by up to 20%

~~ very tentative after this ~~

6. Military and navy?
7. Trade routes?
7b. If ports are needed for trade build more ports?
8. Backfill bureaucracy

### CODE REVIEW

File: EU4toV3/Source/V3World/EconomyManager/EconomyManager.cpp [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+#include "Log.h"
+#include "PoliticalManager/Country/Country.h"
+#include <cmath>
+#include <iomanip>
+#include <numeric>
+#include <ranges>
+
+void V3::EconomyManager::loadCentralizedCountries(const std::map<std::string, std::shared_ptr<Country>>& countries)
+{
+	auto selectCentralized = [](std::shared_ptr<Country> country) {
```
Comment (Zemurin) [inline EU4toV3/Source/V3World/EconomyManager/EconomyManager.cpp:14]: While I always appreciate lovely C++ mannerisms (and Idhrendur does even more, I bet), if I wrote this loop I'd keep it trivial `for (every country) if (country!=decentralized) {shove into vector}`.

I'm not saying change it, but I always keep in mind people with minimal coding knowledge browsing the code. Like me.

**Review by Zemurin — COMMENTED** [review]
Looking good so far!

File: EU4toV3/Source/V3World/PoliticalManager/Country/Country.cpp [hunk 2]
```diff
@@ -41,6 +44,131 @@ void V3::Country::initializeCountry(std::istream& theStream)
...
+		usage += cost * 1; // If we end up mapping institution levels, it is cost * levels
+	}
+
+	// Characters
+	for (const auto& character: processedData.characters)
+	{
+		if (!character.admiral && !character.general)
+			continue;
+
+		if (character.commanderRank == "")
```
Comment (Zemurin) [inline EU4toV3/Source/V3World/PoliticalManager/Country/Country.cpp:103]: I actually don't set commanderRank if it's 0 (shit-tier pips in eu4), since I don't export it. Not all generals have it in vanilla either, I think. Unsure if not setting it defaults it to rank1, which is possible.

File: EU4toV3/Source/V3World/PoliticalManager/Country/Country.cpp [hunk 3]
```diff
@@ -41,6 +44,131 @@ void V3::Country::initializeCountry(std::istream& theStream)
...
+		{
+			Log(LogLevel::Warning) << character.culture << " commander: " << character.lastName << " has no rank.";
+			continue;
+		}
+
+		try
+		{
+			const int rank = std::stoi(character.commanderRank.substr(character.commanderRank.length() - 1));
```
Comment (Zemurin) [inline EU4toV3/Source/V3World/PoliticalManager/Country/Country.cpp:111]: This is untrue, for some reason lvl6 commanders have usage 10 (lvl 5 are correctly usage 30)

File: EU4toV3/Source/V3World/PoliticalManager/Country/Country.cpp [hunk 4]
```diff
@@ -41,6 +44,131 @@ void V3::Country::initializeCountry(std::istream& theStream)
...
+		catch (std::exception& e)
+		{
+			Log(LogLevel::Error) << "Broken military leader rank: " << character.commanderRank << " - " << e.what();
+		}
+	}
+
+	return usage;
+}
+
+bool V3::Country::isTechLocked(const std::vector<std::string>& techs) const
```
Comment (Zemurin) [inline EU4toV3/Source/V3World/PoliticalManager/Country/Country.cpp:123]: Name is unclear. Is what tech locked? Locked how?

File: EU4toV3/Source/V3World/PoliticalManager/PoliticalManager.cpp [hunk 5]
```diff
@@ -334,10 +335,24 @@ void V3::PoliticalManager::setupLaws()
...
+		setupInstitutions(country);
 	}
 	Log(LogLevel::Info) << "<> " << counter << " countries codified.";
 }
 
+void V3::PoliticalManager::setupInstitutions(const std::shared_ptr<Country>& country) const
+{
+	for (const auto& law: country->getProcessedData().laws)
+	{
+		if (auto institution = lawMapper.getLaws().at(law).institution; institution != "")
```
Comment (Zemurin) [inline EU4toV3/Source/V3World/PoliticalManager/PoliticalManager.cpp:349]: `!institution.empty()`

File: EU4toV3Tests/TestFiles/configurables/econ_defines.txt [hunk 6]
```diff
@@ -0,0 +1,5 @@
+global_construction_points = 1451750
+centralized_population_ratio = 0.975
+average_industrial_score = 8
+state_trait_strength = 0.5
+packet_factor = -0.5
```
Comment (Zemurin) [inline EU4toV3Tests/TestFiles/configurables/econ_defines.txt:5]: Maybe this is a good place to define unrecognized industry level, if not using civLevel.

File: EU4toV3Tests/V3WorldTests/EconomyManagerTests/BuildingTests/BuildingGroupsTests.cpp [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+
+	std::stringstream log;
+	std::streambuf* cout_buffer = std::cout.rdbuf();
+	std::cout.rdbuf(log.rdbuf());
+
+	const std::string& parentName = buildingGroups->safeGetParentName("not_a_key0");
+	const int infrastructureCost = buildingGroups->safeGetInfrastructureCost("not_a_key1");
+
+	EXPECT_THAT(log.str(),
+		 testing::HasSubstr(R"( [ERROR] Key not recognized: not_a_key0 is not a recognized building_group. Using an empty group in its place.)"));
```
Comment (Zemurin) [inline EU4toV3Tests/V3WorldTests/EconomyManagerTests/BuildingTests/BuildingGroupsTests.cpp:21]: no need for trailing whitespace before [ERROR]

File: EU4toV3Tests/V3WorldTests/EconomyManagerTests/BuildingTests/BuildingGroupsTests.cpp [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+	V3::BuildingGroupLoader buildingGroupLoader;
+	const auto buildingGroups = buildingGroupLoader.getBuildingGroups();
+	buildingGroupLoader.loadBuildingGroups(modFS);
+
+	auto parentName = buildingGroups->getParentName("not_a_key0");
+	auto infrastructureCost = buildingGroups->getInfrastructureCost("not_a_key1");
+
+	EXPECT_EQ(std::nullopt, parentName);
+	EXPECT_EQ(std::nullopt, infrastructureCost);
+}
```
Comment (Zemurin) [inline EU4toV3Tests/V3WorldTests/EconomyManagerTests/BuildingTests/BuildingGroupsTests.cpp:41]: missing positive test?

File: EU4toV3Tests/V3WorldTests/EconomyManagerTests/NationalBudgetTests/SectorTests.cpp [hunk 9]
```diff
@@ -0,0 +1,39 @@
+#include <gmock/gmock-matchers.h>
+
```
Comment (Zemurin) [inline EU4toV3Tests/V3WorldTests/EconomyManagerTests/NationalBudgetTests/SectorTests.cpp:2]: whitespace

**Review by Zemurin — APPROVED** [review]
Very impressive!

File: EU4ToVic3.sln.DotSettings.user [hunk 10]
```diff
@@ -0,0 +1,6 @@
+﻿<wpf:ResourceDictionary xml:space="preserve" xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml" xmlns:s="clr-namespace:System;assembly=mscorlib" xmlns:ss="urn:shemas-jetbrains-com:settings-storage-xaml" xmlns:wpf="http://schemas.microsoft.com/winfx/2006/xaml/presentation">
```
[the remark this answers is not in the document]
Comment (gawquon) [reply EU4ToVic3.sln.DotSettings.user:1]: Uhh I don't remember setting that setting. Obviously it slipped through in one of my big pushes. I think I accidently clicked some Resharper tooltip that set this up.

File: EU4toV3/Source/V3World/EconomyManager/EconomyManager.cpp [hunk 11]
```diff
@@ -0,0 +1,427 @@
...
+
+		// Use the PM with the most generation available
+		int PMGeneration = 35;
+		if (auto PMName = pickBureaucracyPM(country); PMName)
+		{
+			PMGeneration = PMs.at(PMName.value())->getBureaucracy();
+		}
+
+		// find # of buildings, 100 base value generation
+		const int numAdmins = static_cast<int>((generationTarget - 100) / PMGeneration + 1);
```
[the remark this answers is not in the document]
Comment (gawquon) [reply EU4toV3/Source/V3World/EconomyManager/EconomyManager.cpp:65]: Nope. Not if it's done right. 35 is the minimum possible

File: EU4toV3/Source/V3World/EconomyManager/EconomyManager.cpp [hunk 12]
```diff
@@ -0,0 +1,427 @@
...
+
+		country->distributeGovAdmins(numAdmins);
+	}
+
+	setPMs();
+}
+
+void V3::EconomyManager::hardcodePorts() const
+{
+	for (const auto& country: centralizedCountries)
```
[the remark this answers is not in the document]
Comment (gawquon) [reply EU4toV3/Source/V3World/EconomyManager/EconomyManager.cpp:75]: Nope. Only buildings decentralized get are the automatic Subsistence Farms and Conscription Centers

File: EU4toV3/Source/V3World/EconomyManager/EconomyManager.h [hunk 13]
```diff
@@ -0,0 +1,103 @@
+#ifndef ECONOMY_MANAGER_H
+#define ECONOMY_MANAGER_H
+#include "Configuration.h"
+#include "Date.h"
+#include "Loaders/DefinesLoader/EconDefinesLoader.h"
+#include "PoliticalManager/PoliticalManager.h"
+
+namespace V3
+{
+struct ConfigVariables
```
[the remark this answers is not in the document]
Comment (gawquon) [reply EU4toV3/Source/V3World/EconomyManager/EconomyManager.h:10]: Oh yeah, I forgot to use that struct and just used the parser directly. I'll fix that and yeah rename it.

File: EU4toV3/Source/V3World/Loaders/BuildingLoader/ProductionMethodLoader/ProductionMethodGroupLoader.cpp [hunk 14]
```diff
@@ -0,0 +1,25 @@
...
+		if (getExtension(fileName) == "txt")
+			parseFile(fileName);
+	}
+	clearRegisteredKeywords();
+}
+
+void V3::ProductionMethodGroupLoader::registerKeys()
+{
+	registerRegex(commonItems::catchallRegex, [this](const std::string& PMGroupName, std::istream& theStream) {
+		auto newPMGroup = std::make_unique<V3::ProductionMethodGroup>();
```
[the remark this answers is not in the document]
Comment (gawquon) [reply EU4toV3/Source/V3World/Loaders/BuildingLoader/ProductionMethodLoader/ProductionMethodGroupLoader.cpp:20]: Using make_unique here is basically saying this function is just in charge of the creation of the object. Its a factory/builder like syntax since this is a builder like process. std::move(unique_ptr) to where a shared_ptr is expected turns it into a shared_ptr in the called scope.

File: EU4toV3/Source/V3World/Loaders/NationalBudgetLoader/SectorLoader.cpp [hunk 15]
```diff
@@ -0,0 +1,81 @@
...
+
+	theWeight = mergeValidAdds(country) + theWeight;
+	theWeight = mergeValidMults(country) * theWeight;
+
+	// Severely penalize unrecognized countries for these sectors
+	if (industrial)
+	{
+		if (country.getProcessedData().type != "recognized")
+		{
+			theWeight *= 0.05;
```
[the remark this answers is not in the document]
Comment (gawquon) [reply EU4toV3/Source/V3World/Loaders/NationalBudgetLoader/SectorLoader.cpp:25]: I do need to make better use of CivLevel

File: EU4toV3/Source/V3World/PoliticalManager/Country/Country.cpp [hunk 16]
```diff
@@ -41,6 +44,131 @@ void V3::Country::initializeCountry(std::istream& theStream)
...
+
+	return std::vector<std::shared_ptr<V3::SubState>>{sortedSubstates.begin(), sortedSubstates.begin() + numTopSubstates};
+}
+
+double V3::Country::calculateBureaucracyUsage(const std::map<std::string, V3::Law>& lawsMap) const
+{
+	// None of these hard-coded Vic3 values are in the defines for some reason.
+	double usage = 0.0;
+
+	// Incorporated States 10 * # incorporated
```
[the remark this answers is not in the document]
Comment (gawquon) [reply EU4toV3/Source/V3World/PoliticalManager/Country/Country.cpp:70]: Nope. Also pop only counts if it is in an incorporated state, which I forgot, need to adjust that.

File: EU4toV3/Source/V3World/PoliticalManager/Country/Country.cpp (hunk 3 above)
Comment (gawquon) [reply EU4toV3/Source/V3World/PoliticalManager/Country/Country.cpp:111]: Rank 6 commanders aren't supposed to exist. Unless there is some special case I don't know about.
image

File: EU4toV3/Source/V3World/ClayManager/State/SubState.h [hunk 17]
```diff
@@ -78,10 +82,16 @@ class SubState
 	[[nodiscard]] const auto& getLandshare() const { return landshare; }
 	[[nodiscard]] const auto& getInfrastructure() const { return infrastructure; }
 	[[nodiscard]] const auto& getResource(const std::string& theResource) { return resources[theResource]; }
-	[[nodiscard]] const auto& getTerrainFrequency(const std::string& theTerrain) { return terrainFrequency[theTerrain]; }
+	[[nodiscard]] const auto& getTerrainFrequency(const std::string& terrain) { return terrainFrequency[terrain]; }
```
[the remark this answers is not in the document]
Comment (gawquon) [reply EU4toV3/Source/V3World/ClayManager/State/SubState.h:85]: I thought [key] on mismatch emplaces {key, default_obj_of_type}

File: EU4toV3/Source/V3World/PoliticalManager/Country/Country.cpp [hunk 18]
```diff
@@ -41,6 +44,131 @@ void V3::Country::initializeCountry(std::istream& theStream)
...
+	{
+		if (!lawsMap.contains(law))
+		{
+			Log(LogLevel::Warning) << "Finding bureaucracy multiplier of law: " << law << " that has no definition! Skipping.";
+			continue;
+		}
+
+		lawsMult += lawsMap.at(law).bureaucracyCostMult;
+	}
+	usage += getPopCount() / 25000.0 * (1 + lawsMult);
```
[the remark this answers is not in the document]
Comment (gawquon) [reply EU4toV3/Source/V3World/PoliticalManager/Country/Country.cpp:88]: Right, even if some crazy mod/bug made it so, it is multiply not divide.

[the remark this answers is not in the document]
Comment (Zemurin) [reply EU4toV3/Source/V3World/PoliticalManager/Country/Country.cpp:88]: It will be divide if that ever is -1.

File: EU4toV3/Source/V3World/ClayManager/State/SubState.h (hunk 17 above)
[the remark this answers is not in the document]
Comment (Zemurin) [reply EU4toV3/Source/V3World/ClayManager/State/SubState.h:85]: It very well might. I may have it confused with vector which doesn't do emplacement on [] operator.

File: EU4toV3/Source/V3World/PoliticalManager/Country/Country.h [hunk 19]
```diff
@@ -97,14 +107,20 @@ class Country: commonItems::parser
...
-	void addSubState(const std::shared_ptr<SubState>& subState) { substates.push_back(subState); }
-	void setSubStates(const std::vector<std::shared_ptr<SubState>>& subStates) { substates = subStates; }
+	void addSubState(const std::shared_ptr<SubState>& theSubstate) { substates.push_back(theSubstate); }
+	void setSubStates(const std::vector<std::shared_ptr<SubState>>& theSubstates) { substates = theSubstates; }
 	void setProcessedData(const ProcessedData& data) { processedData = data; }
 
 	[[nodiscard]] std::string getName(const std::string& language) const;
 	[[nodiscard]] std::string getAdjective(const std::string& language) const;
+	[[nodiscard]] int getPopCount() const;
+	[[nodiscard]] static int getPopCount(std::vector<std::shared_ptr<SubState>> theSubstates);
```
Comment (Idhrendur) [inline EU4toV3/Source/V3World/PoliticalManager/Country/Country.h:122]: Add tests for these

File: EU4toV3/Source/V3World/PoliticalManager/Country/Country.cpp [hunk 20]
```diff
@@ -41,6 +43,140 @@ void V3::Country::initializeCountry(std::istream& theStream)
...
+			Log(LogLevel::Warning) << "Finding bureaucracy multiplier of law: " << law << " that has no definition! Skipping.";
+			continue;
+		}
+
+		lawsMult += lawsMap.at(law).bureaucracyCostMult;
+	}
+
+	for (const auto& substate: substates)
+	{
+		if (substate->isIncorporated())
```
Comment (Idhrendur) [inline EU4toV3/Source/V3World/PoliticalManager/Country/Country.cpp:82]: Flip it to 
```
if (!substate->isIncorporated())
{
  continue;
}
```
to reduce the nesting level.

File: EU4toV3/Source/V3World/PoliticalManager/Country/Country.cpp [hunk 21]
```diff
@@ -41,6 +43,140 @@ void V3::Country::initializeCountry(std::istream& theStream)
...
+		{
+			// Incorporated States - 10 per incorporated state
+			usage += 10;
+
+			// Pops - only pops in incorporated states count
+			// Modified by laws - game caps this at 0
+			if (lawsMult > -1)
```
Comment (Idhrendur) [inline EU4toV3/Source/V3World/PoliticalManager/Country/Country.cpp:89]: Better to ensure the multiplier is appropriate above and avoid the logic here or the addition in this calc.

File: EU4toV3/Source/V3World/PoliticalManager/Country/Country.cpp [hunk 22]
```diff
@@ -41,6 +43,140 @@ void V3::Country::initializeCountry(std::istream& theStream)
...
+	};
+
+	std::ranges::sort(sortedSubstates, popComparison);
+
+	const int numTopSubstates = std::max(static_cast<int>(static_cast<double>(sortedSubstates.size()) * percentile), 1);
+
+	return std::vector<std::shared_ptr<V3::SubState>>{sortedSubstates.begin(), sortedSubstates.begin() + numTopSubstates};
+}
+
+double V3::Country::calculateBureaucracyUsage(const std::map<std::string, V3::Law>& lawsMap) const
```
Comment (Idhrendur) [inline EU4toV3/Source/V3World/PoliticalManager/Country/Country.cpp:63]: Split each of the sources of bureaucracy usage out into smaller functions that this function calls in turn.

File: EU4toV3Tests/V3WorldTests/LoaderTests/NationalBudgetLoaderTests/SectorLoaderTests.cpp [hunk 23]
```diff
@@ -0,0 +1,31 @@
+#include "CountryManager/EU4Country.h"
+#include "Loaders/NationalBudgetLoader/SectorLoader.h"
+#include "PoliticalManager/Country/Country.h"
+#include "gtest/gtest.h"
+#include <gmock/gmock-matchers.h>
+
+
```
Comment (Idhrendur) [inline EU4toV3Tests/V3WorldTests/LoaderTests/NationalBudgetLoaderTests/SectorLoaderTests.cpp:7]: Add a test with empty input.

File: EU4toV3/Source/V3World/EconomyManager/NationalBudget/Sector.h [hunk 24]
```diff
@@ -0,0 +1,29 @@
...
+
+namespace V3
+{
+// A Sector of industry is a group of vic3 buildings with similar characteristics and distribution
+class SectorLoader;
+class Country;
+class Sector
+{
+  public:
+	Sector(const std::shared_ptr<SectorLoader>& sectorRules, const std::shared_ptr<Country>& country);
```
Comment (Idhrendur) [inline EU4toV3/Source/V3World/EconomyManager/NationalBudget/Sector.h:13]:
```suggestion
	Sector(const SectorLoader& sectorRules, const Country& country);
```

File: EU4toV3/Source/V3World/EconomyManager/NationalBudget/Sector.cpp [hunk 25]
```diff
@@ -0,0 +1,20 @@
+#include "Sector.h"
+#include "Loaders/NationalBudgetLoader/SectorLoader.h"
+#include "PoliticalManager/Country/Country.h"
+#include <cmath>
+
+V3::Sector::Sector(const std::shared_ptr<SectorLoader>& sectorRules, const std::shared_ptr<Country>& country):
+	 name(sectorRules->getName()), buildings(sectorRules->getBuildings())
```
Comment (Idhrendur) [inline EU4toV3/Source/V3World/EconomyManager/NationalBudget/Sector.cpp:7]:
```suggestion
V3::Sector::Sector(const SectorLoader& sectorRules, const Country& country):
	 name(sectorRules.getName()), buildings(sectorRules.getBuildings())
```

File: EU4toV3/Source/V3World/Loaders/NationalBudgetLoader/NationalBudgetLoader.h [hunk 26]
```diff
@@ -0,0 +1,25 @@
...
+  public:
+	NationalBudgetLoader() = default;
+	void loadNationalBudget(const std::string& filePath);
+
+	[[nodiscard]] const auto& getSectorBlueprints() const { return sectorBlueprints; }
+
+  private:
+	void registerKeys();
+
+	std::vector<std::shared_ptr<SectorLoader>> sectorBlueprints;
```
Comment (Idhrendur) [inline EU4toV3/Source/V3World/Loaders/NationalBudgetLoader/NationalBudgetLoader.h:21]: Shouldn't need to be shared_ptr

File: EU4toV3/Source/V3World/EconomyManager/EconomyManager.cpp [hunk 27]
```diff
@@ -0,0 +1,423 @@
...
+
+		// Use the PM with the most generation available
+		int PMGeneration = 35;
+		if (const auto& PMName = pickBureaucracyPM(country); PMName)
+		{
+			PMGeneration = PMs.at(PMName.value())->getBureaucracy();
+		}
+
+		// find # of buildings, 100 base value generation
+		const int numAdmins = static_cast<int>((generationTarget - 100) / PMGeneration + 1);
```
Comment (Idhrendur) [inline EU4toV3/Source/V3World/EconomyManager/EconomyManager.cpp:65]: Instead of the +1, you could use std::ceil

File: EU4toV3/Source/V3World/EconomyManager/EconomyManager.cpp [hunk 28]
```diff
@@ -0,0 +1,423 @@
...
+			totalIndustryWeight += country->getIndustryWeight();
+			totalIndustryFactor += country->getIndustryFactor();
+		}
+
+		// adjust global total by average industry factor compared to baseline
+		const double globalIndustryFactor = (totalIndustryFactor / static_cast<double>(centralizedCountries.size()) / 0.8) - 1;
+		specialFactors = globalIndustryFactor;
+
+		Log(LogLevel::Info) << std::fixed << std::setprecision(0) << "<> The world is " << (globalIndustryFactor + 1) * 100
+								  << "% industrial compared to baseline. Compensating";
```
Comment (Idhrendur) [inline EU4toV3/Source/V3World/EconomyManager/EconomyManager.cpp:121]: Consider splitting this into a function, especially when you consider the other economyType options

File: EU4toV3/Source/V3World/EconomyManager/EconomyManager.cpp [hunk 29]
```diff
@@ -0,0 +1,423 @@
...
+			subState->setIndustryWeight(std::max(0.0, base * (1 + terrainMultiplier + stateTraitMultiplier)));
+			totalIndustryWeight += subState->getIndustryWeight();
+		}
+
+		// Now back through substates, give each a percentage of country's budget based on percent of total industry score
+		for (const auto& subState: country->getSubStates())
+		{
+			const double stateBudget = country->getCPBudget() * subState->getIndustryWeight() / totalIndustryWeight;
+			subState->setCPBudget(static_cast<int>(std::round(stateBudget)));
+		}
```
Comment (Idhrendur) [inline EU4toV3/Source/V3World/EconomyManager/EconomyManager.cpp:193]: Consider moving all this into a function

File: EU4toV3/Source/V3World/EconomyManager/EconomyManager.cpp [hunk 30]
```diff
@@ -0,0 +1,423 @@
...
+	double terrainMultiplier = 0;
+
+	for (const auto& [terrain, frequency]: subState->getTerrainFrequencies())
+	{
+		if (!stateTerrainModifiers.contains(terrain))
+		{
+			if (terrain == "ocean" || terrain == "lake")
+			{
+				Log(LogLevel::Debug) << "Water terrain found in " << country->getTag() << "'s SubState " << subState->getHomeState()->getName()
+											<< " WTF Paradox. Ignoring.";
```
Comment (Idhrendur) [inline EU4toV3/Source/V3World/EconomyManager/EconomyManager.cpp:325]: The only appropriate response to something we'll no doubt see lots of.

**Review by Idhrendur — APPROVED** [review]

File: EU4toV3/Source/V3World/Loaders/NationalBudgetLoader/Script/MathScript.cpp [hunk 31]
```diff
@@ -0,0 +1,43 @@
...
+{
+	registerKeys();
+	parseStream(theStream);
+	clearRegisteredKeywords();
+}
+
+void V3::MathScript::registerKeys()
+{
+	registerKeyword("value", [this](std::istream& theStream) {
+		value = commonItems::getDouble(theStream);
```
[the remark this answers is not in the document]
Comment (gawquon) [reply EU4toV3/Source/V3World/Loaders/NationalBudgetLoader/Script/MathScript.cpp:18]: Why, what makes this different? It is not std::stod and as far as I have checked, we don't try catch any parser interface commonItems::getDouble() in the previous/this convertor.

[the remark this answers is not in the document]
Comment (Idhrendur) [reply EU4toV3/Source/V3World/Loaders/NationalBudgetLoader/Script/MathScript.cpp:18]: Huh, I must have been tired when reviewing and thought it was a std::stod.
